src/D2L.Bmx/OktaAuthenticator.cs

```diff
@@ -130,6 +136,112 @@ private bool TryAuthenticateFromCache(
+		string? sessionId;
+		string? userEmail;
+
+		try {
+			var page = await browser.NewPageAsync().WaitAsync( cancellationTokenSource.Token );
+			string baseAddress = $"https://{org}.okta.com/";
+			int attempt = 1;
+
+			page.Load += ( _, _ ) => _ = GetSessionCookieAsync( cancellationTokenSource.Token );
+			page.Response += ( _, responseCreatedEventArgs ) => _ = GetOktaUserEmailAsync(
+				responseCreatedEventArgs.Response
+			);
+			await page.GoToAsync( baseAddress, timeout: 10000 );
+			sessionId = await sessionIdTaskProducer.Task.WaitAsync( cancellationTokenSource.Token );
+			userEmail = await userEmailTaskProducer.Task.WaitAsync( cancellationTokenSource.Token );
+
+			async Task GetSessionCookieAsync( CancellationToken cancellationToken ) {
+				var url = new Uri( page.Url );
+				if( url.Host == $"{org}.okta.com" ) {
+					string title = await page.GetTitleAsync();
+					if( title.Contains( "sign in", StringComparison.OrdinalIgnoreCase ) ) {
+						if( attempt < 3 && url.AbsolutePath != "/" ) {
+							attempt++;
+							await page.GoToAsync( baseAddress );
+						} else {
+							sessionIdTaskProducer.SetResult( null );
+						}
+						return;
+					}
+				}
+				var cookies = await page.GetCookiesAsync( baseAddress );
+				if( Array.Find( cookies, c => c.Name == "sid" )?.Value is string sid ) {
+					sessionIdTaskProducer.SetResult( sid );
+				}
+			}
+
+			async Task GetOktaUserEmailAsync(
+				IResponse response
+			) {
+				if( response.Url.Contains( $"{baseAddress}enduser/api/v1/home" ) ) {
+					string content = await response.TextAsync();
+					var home = JsonSerializer.Deserialize( content, JsonCamelCaseContext.Default.OktaHomeResponse );
+					if( home is not null ) {
+						userEmailTaskProducer.SetResult( home.Login );
+					}
+				}
+			}
+
+		} catch( TaskCanceledException ) {
+			consoleWriter.WriteWarning(
+				$"WARNING: Failed to create {org} Okta session through DSSO. Check if org is correct."
+					+ " If running BMX with admin privileges, rerun the command with the '--experimental' flag."
+			);
+			return null;
+		} catch( TargetClosedException ) {
+			consoleWriter.WriteWarning(
+				$"WARNING: Failed to create {org} Okta session through DSSO as BMX is likely being run with elevated privilieges." +
+					" Rerun the command with the '--experimental' flag."
+			);
+			return null;
+		} catch( Exception ) {
+			consoleWriter.WriteWarning( "WARNING: Unknown error while trying to authenticate with Okta using DSSO." );
+			return null;
+		} finally {
+			cancellationTokenSource.Dispose();
+			browser.Dispose();
+		}
+
+		if( sessionId is null || userEmail is null ) {
+			return null;
+		} else if( !OktaUserMatchesProvided( userEmail, user ) ) {
+			consoleWriter.WriteWarning(
+				"WARNING: Could not create Okta session using DSSO as "
+				+ $"provided Okta user '{user}' does not match user '{userEmail}'." );
```

Comment:
Potentially confusing if `user` doesn't have the `@desire2learn.com` included even though we are not including that in our comparison.